File: tests/java/tests-java-gradle-junit5-allure3-retrofit2/src/test/java/tests/api/AuthRoundTripApiTests.java
package tests.api;

import annotations.Layer;
import api.ApiTestBase;
import api.Calls;
import api.model.LoginRequest;
import api.model.RegisterRequest;
import helpers.DataFaker;
import io.qameta.allure.Epic;
import io.qameta.allure.Feature;
import io.qameta.allure.Severity;
import io.qameta.allure.SeverityLevel;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

@Layer("api")
@Epic("Authentication")
@Feature("Account lifecycle")
@Severity(SeverityLevel.CRITICAL)
@DisplayName("Auth account lifecycle on deployed stand")
class AuthRoundTripApiTests extends ApiTestBase {

    /**
     * Full account lifecycle across separate HTTP requests — proves DB and JWT are wired
     * together on the deployed stand, and documents that logout is stateless: the JWT keeps
     * working until the account itself is gone. Deletes the user it registers, so the stand
     * does not accumulate test accounts.
     */
    @Test
    @Tag("api")
    @DisplayName("register → login → me → logout (stateless: token survives) → delete → me is 401")
    void accountLifecycleRoundTrip() {
        String username = DataFaker.username();
        String password = "password123";

        var registered = Calls.execute(api.register(new RegisterRequest(username, password)));
        assertEquals(201, registered.code());
        assertNotNull(registered.body());
        assertEquals(username, registered.body().username());

        var loggedIn = Calls.execute(api.login(new LoginRequest(username, password)));
        assertEquals(200, loggedIn.code());
        assertNotNull(loggedIn.body());
        String token = loggedIn.body().token();

        var me = Calls.execute(api.me(bearer(token)));
        assertEquals(200, me.code());
        assertNotNull(me.body());
        assertEquals(username, me.body().username());

        assertEquals(204, Calls.execute(api.logout(bearer(token))).code());

        // Stateless JWT: logout does not invalidate the token server-side — by design.
        var meAfterLogout = Calls.execute(api.me(bearer(token)));
        assertEquals(200, meAfterLogout.code());
        assertNotNull(meAfterLogout.body());
        assertEquals(username, meAfterLogout.body().username());

        assertEquals(204, Calls.execute(api.deleteMe(bearer(token))).code());

        // The token still verifies cryptographically, but the account is gone → 401.
        assertEquals(401, Calls.execute(api.me(bearer(token))).code());
    }
}

File: tests/csharp/tests-csharp-xunit-api_request-playwright/Tests/Api/AuthRoundTripApiTests.cs
using Allure.Net.Commons;
using Allure.Net.Commons.Attributes;
using Api;
using Api.Models;
using Helpers;
using Tests;

namespace Tests.Api;

[AllureEpic("Authentication")]
[AllureFeature("Account lifecycle")]
[AllureSeverity(SeverityLevel.critical)]
[AllureSuite("Auth account lifecycle on deployed stand")]
public sealed class AuthRoundTripApiTests : ApiTestBase
{
    /// <summary>
    /// Full account lifecycle across separate HTTP requests — proves DB and JWT are wired
    /// together on the deployed stand, and documents that logout is stateless: the JWT keeps
    /// working until the account itself is gone. Deletes the user it registers, so the stand
    /// does not accumulate test accounts.
    /// </summary>
    [Trait("TestCategory", "api")]
    [Fact(DisplayName = "register → login → me → logout (stateless: token survives) → delete → me is 401")]
    public void AccountLifecycleRoundTrip()
    {
        var username = DataFaker.Username();
        const string password = "password123";

        var registered = PlaywrightHttp.Request(
            "POST",
            "/api/auth/register",
            json: new RegisterRequest(username, password));
        Assert.True(Equals(201, registered.Status), registered.Body);
        Assert.Equal(username, registered.Text("username"));

        var login = PlaywrightHttp.Request(
            "POST",
            "/api/auth/login",
            json: new LoginRequest(username, password));
        Assert.True(Equals(200, login.Status), login.Body);
        var token = login.Text("token");

        var me = PlaywrightHttp.Request("GET", "/api/auth/me", token: token);
        Assert.True(Equals(200, me.Status), me.Body);
        Assert.Equal(username, me.Text("username"));

        var logout = PlaywrightHttp.Request("POST", "/api/auth/logout", token: token);
        Assert.True(Equals(204, logout.Status), logout.Body);

        // Stateless JWT: logout does not invalidate the token server-side — by design.
        var meAfterLogout = PlaywrightHttp.Request("GET", "/api/auth/me", token: token);
        Assert.True(Equals(200, meAfterLogout.Status), meAfterLogout.Body);
        Assert.Equal(username, meAfterLogout.Text("username"));

        var deleted = PlaywrightHttp.Request("DELETE", "/api/auth/me", token: token);
        Assert.True(Equals(204, deleted.Status), deleted.Body);

        // The token still verifies cryptographically, but the account is gone → 401.
        var meAfterDelete = PlaywrightHttp.Request("GET", "/api/auth/me", token: token);
        Assert.True(Equals(401, meAfterDelete.Status), meAfterDelete.Body);
    }
}
